PD-783: Makes Nav responsive

brookescarlett opened the pull request:
## ✍️ Proposed changes

- Makes Nav responses
- Uses styled API for GridContainer and migrates GridItem off of imports from @emotion/core

🎟 _Jira ticket:_ [PD-783](https://jira.mongodb.org/browse/PD-783)

## 🛠 Types of changes

<!--
What types of changes does your code introduce? Put an `x` in the applicable boxes.
-->

- [ ] Tooling (updates to workspace config or internal tooling)
- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

![Screen Shot 2020-09-29 at 5 26 45 PM](https://user-images.githubusercontent.com/26016393/94618513-495c0d80-0279-11eb-89c6-05546afe6b4e.png)

lazytype commented on website/components/grid/Grid.tsx:
```diff
@@ -59,23 +52,22 @@ type GridContainerProps = JSX.IntrinsicElements['div'] & {
-
-  return <div css={gridContainerStyles}>{children}</div>;
-}
+const GridContainer = styled<'div', GridContainerProps>('div')`
+  ${mq({
+    width: breakpointGaps.map(gap => `calc(100% + ${gap})`),
+    margin: breakpointGaps.map(gap => `calc(-1 * ${gap} / 2)`),
+  })}
+
+  display: flex;
+  ${['flex-direction']}: ${props =>
+    props.direction ? props.direction : Direction.Row};
+  ${['flex-wrap']}: ${props => (props.wrap ? props.wrap : Wrap.Wrap)};
+  ${['align-items']}: ${props => (props.align ? props.align : Align.Center)};
+  ${['justify-content']}: ${props =>
+    props.justify ? props.justify : Justify.Center};
+`;
```
not a big fan of this api. it looks more like its own language than javascript and is significantly harder to read than the previous code.

brookescarlett replied:
Updated because @DesignerDave loves styled components. Let me know what you think in this case!

lazytype replied:
I'd have to hear a real rationale behind the preference. the readability of this code does not make a very convincing case.

lazytype commented on website/components/navigation/MobileNavigationGroup.tsx:
```diff
@@ -0,0 +1,134 @@
+
+  return (
+    <li
+      className={css`
```
we probably want to pull out `className` from `rest` and merge it in here

brookescarlett replied:
👍

lazytype commented on website/components/navigation/MobileNavigationGroup.tsx:
```diff
@@ -0,0 +1,134 @@
+          <div
+            ref={nodeRef}
+            className={cx(defaultStyle, {
+              [transitionStyles.entering]: state === 'entering',
+              [css`
+                opacity: 1;
+                max-height: ${ulRef?.current?.getBoundingClientRect().height}px;
+              `]: state === 'entered',
+              [transitionStyles.exiting]: state === 'exiting',
+              [transitionStyles.exited]: state === 'exited',
```
this could be simplified to `cx(defaultStyle, transitionStyles[state], ...)`

brookescarlett replied:
I can't only because in the style for entered I need access to `ulRef`, although I suppose I can make entered a function and just pass the value 🤔

lazytype replied:
that definition can stay as it is, e.g.
```ts
cx(defaultStyle, transitionStyles[state], {
  [css`
    opacity: 1;
    max-height: ${ulRef?.current?.getBoundingClientRect().height}px;
  `]: state === 'entered'
}
```

brookescarlett replied:
👍

lazytype commented on website/components/navigation/Navigation.tsx:
```diff
@@ -19,57 +23,73 @@ const logoStyles = css`
 
 function Navigation() {
+  const viewport = useViewportSize();
+  const isMobile = viewport && viewport.width < 768;
```
`!!viewport` to keep `isMobile` to a boolean

brookescarlett replied:
👍

lazytype commented on website/components/navigation/Navigation.tsx:
```diff
@@ -19,57 +23,73 @@ const logoStyles = css`
+  const Group = isMobile ? MobileNavigationGroup : SideNavGroup;
+  const Item = isMobile ? MobileNavigationItem : SideNavItem;
+
+  const groupProps =
```
prefer using ternary operator here so that `groupProps` can't be `false`

brookescarlett replied:
👍

lazytype commented on website/components/navigation/Navigation.tsx:
```diff
@@ -19,57 +23,73 @@ const logoStyles = css`
+    </>
+  );
+
+  if (isMobile) {
```
this definition has a lot of branching based on `isMobile` which makes it harder to reason about. I would explore what it would look like if there was only one branching point.

brookescarlett replied:
The main thing I wanted to do here was share the the `content` between the Mobile and desktop Navigation components.

brookescarlett replied:
Is it easier to make sense of if I move all of the conditional logic to a component, so it only reads as such:

```
function Navigation() {
  const viewport = useViewportSize();
  const isMobile = !!viewport && viewport.width < 768;

  if (isMobile) {
    return (
      <MobileNavigation>
        <Content isMobile />
      </MobileNavigation>
    );
  }

  return (
    <nav className={navWidth}>
      <MDBDesignLogo className={logoStyles} />
      <SideNav>
        <Content />
      </SideNav>
    </nav>
  );
}
```

lazytype replied:
yeah, that could be an improvement

lazytype replied:
oh I see you already made the change. 👍 LGTM

brookescarlett replied:
👍

lazytype commented on website/components/navigation/styles.ts:
```diff
@@ -0,0 +1,18 @@
+export const ulStyleOverrides = css`
+  margin-block-start: 0px;
+  margin-block-end: 0px;
+  padding-inline-start: 0px;
```
saw various places we could replace `0px` with `0` in this PR

brookescarlett replied:
Fixed

brookescarlett commented:
@DesignerDave: updated so that you can collapse the nav (similar to a modal, collapses when you click the background per convo with @audroue). Also added routing to this PR. Made a separate ticket to better address the header (because it needs to be more responsive: https://jira.mongodb.org/browse/PD-915).

DesignerDave reviewed (approved):
I think the animation needs a bit of work, but no need to block this for the time being!
